File: golang/from_snippets/BedmasParser/Bedmas4.go
// You can edit this code!
// Click here and start typing.
package main

import "fmt"
import "strings"
import "strconv"
import "math"

type Entry struct {
	op rune // operation, rune is like char
	num float64 // number
}

// Pretty printing
func (e Entry) String() string {
	if e.op != 0 {
		return fmt.Sprintf("%v", string(e.op))
	}else{
		return fmt.Sprintf("%v", e.num)
	}
}

func main() {
	
	str := "2234*((1+2)^2 + (3-5)*2)"
	arg := []Entry{} // slice (resizeable array) of Entries
	p1, p2 := 0, 0 // position 1 and 2
	var result float64
	
	// Make Entries from the input string
	for ;; {

		// From p1, find the next operation character, set that position to p2
		p2 = strings.IndexAny(str[p1:], "*/+-^()" )
		if p2 == -1 { break }

		// If there's a number p1 and p2 then append it
		float, err := strconv.ParseFloat(strings.TrimSpace(str[p1:p1+p2]),64)
		if err == nil {
			e1 := Entry {num: float}
			arg = append(arg, e1)
		}
		// Append the operation
		e2 := Entry {op: rune(str[p1+p2])}
		arg = append(arg, e2)

		p1=p1+p2+1
	}
	
	// Print Entries
	fmt.Println(arg)
			
	for ;len(arg) > 1; {
	
		// Find deepest paren
		start := 0
		end := len(arg)
		
		for i, v := range arg {
			if v.op == ')' { end = i; break }
		}
		for i := end-1;i>=0; i-=1 {
			if arg[i].op == '(' { start = i; break }
		}
	
		// Find the position of the highest priority operation
		pos := 0
		
		for _, char := range "^*/+-" { 
			for i, v := range arg[start:end] {
				if v.op == rune(char) { pos = start+i; break }
			}
			if pos != 0 { break }
		}

		// Find result
		a := arg[pos-1].num
		b := arg[pos+1].num
		
		switch arg[pos].op {
		case '^': 
			result = math.Pow(a,b)
		case '*': 
			result = a*b
		case '/': 
			result = a/b
		case '+': 
			result = a+b
		case '-': 
			result = a-b
		default:
			panic("argh...")
		}

		// Exit if there's only three entries at this stage
		if len(arg) <= 3 { break }
	

		// Otherwise reform argument
		newArg := []Entry{}
		newEntry := Entry{num: result}
		
		if arg[pos-2].op =='(' && arg[pos+2].op == ')' {	
			newArg = append(newArg, arg[:pos-2]...)
			newArg = append(newArg, newEntry)
			newArg = append(newArg, arg[pos+3:]...)
		} else {
			newArg = append(newArg, arg[:pos-1]...)
			newArg = append(newArg, newEntry)
			newArg = append(newArg, arg[pos+2:]...)

		}
		arg = newArg
		
		// Debugging
		fmt.Println(arg)
	}
	// debugging
	fmt.Println(result)
}

File: golang/from_snippets/BedmasParser/Bedmas4.r
rebol []

split-any: func [ str chars2 /local out p1 p2 chars ] [

	; Split a string at any of the given chars in the second argument
	; split-any "2234*((1+2)^^2 + (3-5)*2)" "*/+-^^()e<>="
	; == ["2234" #"*" #"(" #"(" "1" #"+" "2" #")" #"^^" "2 " #"+" " " #"(" "3" #"-" "5" #")" #"*" "2" #")"]

	p1: p2: str ; position 1 and 2
	out: copy []
	chars: charset chars2

	while [ p2: find p1 chars ] [

		; probe p1 probe p2 probe out input

		if p1 <> p2 [
			append out copy/part p1 p2
		]
		append out p2/1
		p1: p2: next p2

	]
	; append the rest
	if not empty? p1 [ append out p1 ]
	out
]



do-math: func [ arg  /debug /local start end pos result] [

	{ do-math "2234 * ( ( 1 + 2 ) ^^ 2 + ( 3 - 5 ) * 2 )"
	"2234 * ( 3 ^^ 2 + ( 3 - 5 ) * 2 )"
	"2234 * ( 3 ^^ 2 + -2 * 2 )"
	"2234 * ( 9.0 + -2 * 2 )"
	"2234 * ( 9.0 + -4 )"
	"2234 * 5.0"
	"11170.0"
	== 11170.0 }
	
	; remove whitespace
	remove-each a arg [ find { ^-^/} a ] 
	
	; split string into operations and numbers
	arg: split-any arg "*/+-^^()e<>=" 

	while [ 1 <> length? arg ] [

		; range of interest is either parens (parentheses) or the whole expression
		start: arg
		end: tail arg
	
		; set range to parens, if there are any
		if pos: find arg #")" [
			start: pos
			end: find/reverse pos #"("
		]
		; find deepest paren by finding the first closing paren then searching backwards 
		; to its opening paren
	
		; find the highest priority operation - num1 op num2
		; ^ is a rebol escape char so ^^ just means ^
		foreach char "^^*/+-" [
			if pos: attempt [ find/part start char end ] [break]
		]
		pos: back pos ; go to num1
		
		
		; evalute binomial operation
		either #"^^" = pos/2 [ 
			result: power do pos/1 do pos/3
		] [
			result: do reform copy/part pos 3
		]
	
		; if the parenthesis only contains three values then remove the parens as well
		either all [ 
			#"(" = pos/-1
			#")" = pos/4
		] [
			change/part back pos result 5
		] [
			change/part pos result 3
		]
		
		if debug [
			probe reform arg
			input
		]
	]
	return arg/1
]

; do-math/debug "2234*((1+2)^^2 + (3-5)*2)" 
do-math "2234 * ( ( 1 + 2 ) ^^ 2 + ( 3 - 5 ) * 2 )"
; took 269 microseconds
; 0:00:00.000269
; roughly 10X the time of the golang version
